static_website/dispatch.ml

```diff
@@ -1,53 +1,50 @@
 open Lwt.Infix
-open V1_LWT
 
-module Main (C:CONSOLE) (FS:KV_RO) (S:Cohttp_lwt.Server) = struct
+let server_src = Logs.Src.create "server" ~doc:"HTTP server"
+module Server_log = (val Logs.src_log server_src : Logs.LOG)
 
-  let start c fs http =
+module Main (Clock:V1.CLOCK) (FS:V1_LWT.KV_RO) (S:Cohttp_lwt.Server) = struct
+  module Logs_reporter = Mirage_logs.Make(Clock)
 
-    let read_fs name =
-      FS.size fs name >>= function
+  let read_fs fs name =
+    FS.size fs name >>= function
+    | `Error (FS.Unknown_key _) -> Lwt.fail (Failure ("read " ^ name))
+    | `Ok size ->
+      FS.read fs name 0 (Int64.to_int size) >>= function
       | `Error (FS.Unknown_key _) -> Lwt.fail (Failure ("read " ^ name))
-      | `Ok size ->
-        FS.read fs name 0 (Int64.to_int size) >>= function
-        | `Error (FS.Unknown_key _) -> Lwt.fail (Failure ("read " ^ name))
-        | `Ok bufs -> Lwt.return (Cstruct.copyv bufs)
-    in
+      | `Ok bufs -> Lwt.return (Cstruct.copyv bufs)
 
-    (* Split a URI into a list of path segments *)
-    let split_path uri =
-      let path = Uri.path uri in
-      let rec aux = function
-        | [] | [""] -> []
-        | hd::tl -> hd :: aux tl
-      in
-      List.filter (fun e -> e <> "")
-        (aux (Re_str.(split_delim (regexp_string "/") path)))
-    in
+  let rec dispatcher fs uri =
+    match Uri.path uri with
+    | ("" | "/") as path ->
+      Server_log.info (fun f -> f "request for '%s'" path);
+      dispatcher fs (Uri.with_path uri "index.html")
+    | path ->
+      Server_log.info (fun f -> f "request for '%s'" path);
+      Lwt.catch (fun () ->
+          read_fs fs path >>= fun body ->
+          let mime_type = Magic_mime.lookup path in
+          let headers = Cohttp.Header.init () in
+          let headers = Cohttp.Header.add headers "content-type" mime_type in
+          S.respond_string ~status:`OK ~body ~headers ()
+        ) (fun _exn -> S.respond_not_found ())
```

Comment:
This would be a good place to log the error. Also, it should probably return a 500 error code rather than not found, which isn't an exceptional case anyway.